### PR TITLE
feat: reject market-updates that modify enabled field in ante-handler

### PR DESCRIPTION
**Pull request by nivasan1**
## In This PR
- I modify the `ValidateMarketUpdateDecorator`'s `AnteHandler` to reject `MarketMap` market updates that mutate markets as follows
1. For markets that do not exist in the market-map / have yet to be added to the `PricesKeeper`
      - These markets should always be `Disabled`, any update that results in a market-map violating this will be rejected.
2. For markets that have already been added to the `PricesKeeper`
     - The enabled field of the market in the market-map should never be changed by incoming market-updates

## Summary by CodeRabbit

- **New Features**
	- Introduced a `MarketMapKeeper` to enhance market data management in the application.
	- Improved validation of market updates, ensuring proper handling of enabled and disabled market states.
	- Added new error constants for better feedback on market update scenarios.

- **Bug Fixes**
	- Refined logic for market updates, preventing issues with enabling previously disabled markets.

- **Tests**
	- Expanded test cases to cover various market states and ensure robust validation of market updates, including conditions for handling enabled and disabled markets.

### CODE REVIEW

File: protocol/app/ante/market_update_test.go [hunk 1]
```diff
@@ -425,18 +443,117 @@ func TestValidateMarketUpdateDecorator_AnteHandle(t *testing.T) {
...
+				_, err := tApp.App.PricesKeeper.CreateMarket(ctx, mp, prices_types.MarketPrice{
+					Id:       marketID,
+					Exponent: -8,
+					Price:    10,
+				})
+				require.NoError(t, err)
+			}
+
+			// setup initial market-map markets
+			for _, market := range tt.args.marketMapMarkets {
```
Comment (chenyaoy) [inline protocol/app/ante/market_update_test.go:529]: Can you move this to before creating the market in x/prices? I'm adding a dependency when creating the market in x/prices that the market needs to exist in market map

File: protocol/app/ante/market_update_test.go [hunk 2]
```diff
@@ -425,18 +443,117 @@ func TestValidateMarketUpdateDecorator_AnteHandle(t *testing.T) {
...
+			args: args{
+				msgs: []sdk.Msg{
+					&mmtypes.MsgUpsertMarkets{
+						Authority: constants.BobAccAddress.String(),
+						Markets: []mmtypes.Market{
+							testMarketWithEnabled,
+						},
+					},
+				},
+				simulate: true,
```
Comment (chenyaoy) [inline protocol/app/ante/market_update_test.go:473]: should this have `marketMapMarkets: []mmtypes.Market{testMarket}, // existing mm market is disabled` to match the case above (with different `simulate` values)?

File: protocol/app/ante/market_update_test.go [hunk 3]
```diff
@@ -425,18 +443,117 @@ func TestValidateMarketUpdateDecorator_AnteHandle(t *testing.T) {
...
+							testMarket,
+						},
+					},
+				},
+				simulate:         true,
+				marketMapMarkets: []mmtypes.Market{testMarketWithEnabled},
+				marketParams:     []prices_types.MarketParam{testMarketParams},
+			},
+			wantErr: true,
+		},
```
Comment (chenyaoy) [inline protocol/app/ante/market_update_test.go:510]: add tests for MsgUpdateMarkets as well?

File: protocol/app/ante/market_update.go [hunk 4]
```diff
@@ -113,6 +128,73 @@ func (d ValidateMarketUpdateDecorator) doMarketsContainCrossMarket(ctx sdk.Conte
...
+		return err
+	}
+
+	// check validity of incoming market-updates
+	for _, market := range markets {
+		_, exists := mpMap[market.Ticker.CurrencyPair.String()]
+		if !exists {
+			// if market does not exist in x/prices, it should be disabled
+			if market.Ticker.Enabled {
+				return errors.New("newly added market should be disabled")
```
Comment (vincentwschau) [inline protocol/app/ante/market_update.go:154]: nit: Can we add errors registered to the x/prices module for these?

File: protocol/app/ante/market_update_test.go [hunk 5]
```diff
@@ -425,18 +443,197 @@ func TestValidateMarketUpdateDecorator_AnteHandle(t *testing.T) {
...
 	}
```
Comment (chenyaoy) [inline protocol/app/ante/market_update_test.go:591]: Can you add test to successfully update/upsert market and change a field that's not Enabled?

**Review by vincentwschau — APPROVED** [review]

**Review by chenyaoy — APPROVED** [review]

**Pull request merged**